core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/FPCInvokerReactive.scala

```diff
@@ -0,0 +1,483 @@
+
+  private val etcdWorkerFactory =
+    (f: ActorRefFactory) => f.actorOf(EtcdWorker.props(etcdClient, leaseService))
+
+  val dataManagementService: ActorRef =
+    actorSystem.actorOf(DataManagementService.props(watcherService, etcdWorkerFactory))
+
+  private val warmedSchedulers = TrieMap[String, String]()
+  private var warmUpWatcher: Option[Watch] = None
+
+  /**
+   * Factory used by the ContainerProxy to physically create a new container.
+   *
+   * Create and initialize the container factory before kicking off any other
+   * task or actor because further operation does not make sense if something
+   * goes wrong here. Initialization will throw an exception upon failure.
+   */
+  private val containerFactory =
+    SpiLoader
+      .get[ContainerFactoryProvider]
+      .instance(
+        actorSystem,
+        logging,
+        config,
+        instance,
+        Map(
+          "--cap-drop" -> Set("NET_RAW", "NET_ADMIN"),
+          "--ulimit" -> Set("nofile=1024:1024"),
+          "--pids-limit" -> Set("1024")) ++ logsProvider.containerParameters)
+  containerFactory.init()
+
+  CoordinatedShutdown(actorSystem)
+    .addTask(CoordinatedShutdown.PhaseBeforeActorSystemTerminate, "cleanup runtime containers") { () =>
+      containerFactory.cleanup()
+      Future.successful(Done)
+    }
+
+  /** Initialize needed databases */
+  private val entityStore = WhiskEntityStore.datastore()
+  private val activationStore =
+    SpiLoader.get[ActivationStoreProvider].instance(actorSystem, logging)
+
+  private val authStore = WhiskAuthStore.datastore()
+
+  private val namespaceBlacklist: NamespaceBlacklist = new NamespaceBlacklist(authStore)
+
+  Scheduler.scheduleWaitAtMost(loadConfigOrThrow[NamespaceBlacklistConfig](ConfigKeys.blacklist).pollInterval) { () =>
+    logging.debug(this, "running background job to update blacklist")
+    namespaceBlacklist.refreshBlacklist()(ec, TransactionId.invoker).andThen {
+      case Success(set) => logging.info(this, s"updated blacklist to ${set.size} entries")
+      case Failure(t)   => logging.error(this, s"error on updating the blacklist: ${t.getMessage}")
+    }
+  }
+
+  val containerProxyTimeoutConfig = loadConfigOrThrow[ContainerProxyTimeoutConfig](ConfigKeys.containerProxyTimeouts)
+
+  private def getWarmedContainerLimit(invocationNamespace: String): Future[(Int, FiniteDuration)] = {
+    implicit val trasnid = TransactionId.unknown
+    Identity
+      .get(authStore, EntityName(invocationNamespace))(trasnid)
+      .map { identity =>
+        val warmedContainerKeepingCount = identity.limits.warmedContainerKeepingCount.getOrElse(1)
+        val warmedContainerKeepingTimeout = Try {
+          identity.limits.warmedContainerKeepingTimeout.map(Duration(_).toSeconds.seconds).get
+        }.getOrElse(containerProxyTimeoutConfig.keepingDuration)
+        (warmedContainerKeepingCount, warmedContainerKeepingTimeout)
```

Comment:
`warmedContainerKeepingCount, warmedContainerKeepingTimeout` is for `avoid cold start`, doesn't remove one namespace's all warmed containers when container timeout happens, keeps the right amount warmed containers running for some time, when invocation happens next time, have no need to create a container, just uses the non-deleted warmed container.

for a specified namespace, we can configure the warmedContainerKeepingCount, warmedContainerKeepingTimeout in  subjects's limit doucment, e.g.
![image](https://user-images.githubusercontent.com/11749867/121156749-9027eb80-c87b-11eb-81ba-128cc1f70cb8.png)

If doesn't configure `warmedContainerKeepingCount, warmedContainerKeepingTimeout` in couchdb, it will use default configuration (warmedContainerKeepingCount: 1, warmedContainerKeepingTimeout: "60 minutes")

Reply:
how does this work for individual functions? If it's by namespace you can get into situations where one function starves all of the warm containers kept for that namespace right?

Reply:
Regarding `get into situations where one function starves all of the warm containers kept for that namespace right?`
Let's assume warmedContainerKeepCount is `2`, and same namespace has 4 actions (e.g. actionA, actionB, actionC, actionD)
Normally, these 4 actions will complete for `warmedContainerKeepCount: 2`,
* if 1 or 2 actions are executed actions frequently with medium tps
  these actions's 2 warmed containers will keep `warmedContainerkeepingTimeout` before removed even if no activations comes anymore.
* if > 2 actions are executed actions frequently with medium tps
  Just satisfy 2 actions's exist non-removed containers, another 2 actions should do `cold start`, for this situation, use can configure warmedContainerKeepCount with a little bigger value, decide by user.
* big tps for actionA
It seems actionA will use all  `warmedContainerKeepCount : 2` warmed container, other actions(e.g. actionB, actionC, actionD) will do `cold start`, seems have no method to solve this, due to even if we configure a little bigger value, actionA will use all due to actionA's tps is big.

Currently, we just configure it for namespace not for ` individual functions`